urlb5d5be4365_abmargb_oursim_ext_tgz-pJ8-br_edu_ufcg_lsd_oursim_experiments_RemoteNoFScenarioGeneratorJ8/src/br/edu/ufcg/lsd/oursim/events/peer/accounting/RemoteCPUAccountingStrategy.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (C) 2008 Universidade Federal de Campina Grande
+ *  
+ * This file is part of OurGrid. 
+ *
+ * OurGrid is free software: you can redistribute it and/or modify it under the
+ * terms of the GNU Lesser General Public License as published by the Free 
+ * Software Foundation, either version 3 of the License, or (at your option) 
+ * any later version. 
+ * 
+ * This program is distributed in the hope that it will be useful, but WITHOUT 
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
+ * for more details. 
+ * 
+ * You should have received a copy of the GNU Lesser General Public License 
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ * 
+ */
+package br.edu.ufcg.lsd.oursim.events.peer.accounting;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import br.edu.ufcg.lsd.oursim.entities.accounting.ReplicaAccounting;
+import br.edu.ufcg.lsd.oursim.entities.job.ExecutionState;
+
+public class RemoteCPUAccountingStrategy implements AccountingStrategy {
+
+	/* (non-Javadoc)
+	 * @see org.ourgrid.refactoring.peer.controller.accounting.AccountingStrategy#evaluate(java.util.Map)
+	 */
+	public Map<String, Double> evaluate(
+			Map<String, List<ReplicaAccounting>> processes) {
+		
+		Map<String, Double> cpuBalances = new HashMap<String, Double>();
+		
+		double totalCpuTime = 0.;
+		int receivedFavoursCount = 0;
+		
+		for (String providerDN : processes.keySet()) {
+			double totalCpuPerPeer = 0.;
+			
+			double finishedReplicaCpuPerPeer = 0.;
+			int finishedReplicasCount = 0;
+			
+			for (ReplicaAccounting process : processes.get(providerDN)) {
+				Double cpuTime = (double) process.getCpuTime();
+				totalCpuPerPeer += cpuTime;
+				
+				if (process.getState().equals(ExecutionState.FINISHED)) {
+					finishedReplicaCpuPerPeer += cpuTime;
+					finishedReplicasCount++;
+				}
+				
+			}
+			
+			totalCpuTime += finishedReplicaCpuPerPeer;
+			receivedFavoursCount += finishedReplicasCount;
+			
+			double averageCpuPerPeer = totalCpuPerPeer/(finishedReplicaCpuPerPeer/finishedReplicasCount);
+			cpuBalances.put(providerDN, averageCpuPerPeer);
+		}
+		
+		double totalAverage = totalCpuTime / receivedFavoursCount;
+		
+		for (String providerDN : cpuBalances.keySet()) {
+			cpuBalances.put(providerDN, cpuBalances.get(providerDN) * totalAverage);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.